client/lib/sentry/index.ts

```diff
@@ -0,0 +1,211 @@
+		// @ts-expect-error We have a union of tuples and TypeScript wants a Tuple. It's OK.
+		state.sentry[ method ]( ...args );
+		return;
+	}
+	if ( status === 'error' || status === 'disabled' ) {
+		return;
+	}
+	callQueue.push( { f: method, a: args } );
+}
+export function addBreadcrumb( ...args: Parameters< typeof SentryApi.addBreadcrumb > ) {
+	dispatchSentryMethodCall( 'addBreadcrumb', args );
+}
+export function captureEvent( ...args: Parameters< typeof SentryApi.captureEvent > ) {
+	dispatchSentryMethodCall( 'captureEvent', args );
+}
+export function captureException( ...args: Parameters< typeof SentryApi.captureException > ) {
+	dispatchSentryMethodCall( 'captureException', args );
+}
+export function captureMessage( ...args: Parameters< typeof SentryApi.captureMessage > ) {
+	dispatchSentryMethodCall( 'captureMessage', args );
+}
+export function configureScope( ...args: Parameters< typeof SentryApi.configureScope > ) {
+	dispatchSentryMethodCall( 'configureScope', args );
+}
+export function withScope( ...args: Parameters< typeof SentryApi.withScope > ) {
+	dispatchSentryMethodCall( 'withScope', args );
+}
+
+// Replays all calls to the Sentry API
+function processErrorQueue() {
+	for ( const call of callQueue ) {
+		// @ts-expect-error We have a union of tuples and TypeScript wants a Tuple. It's OK.
+		state.sentry[ call.f ]( ...call.a );
+	}
+
+	// And now capture all previously caught exceptions
+	// Because we installed the SDK, at this point we have an access to TraceKit's handler,
+	// which can take care of browser differences (eg. missing exception argument in onerror)
+	if ( window.onerror ) {
+		for ( const error of errorQueue ) {
+			window.onerror( ...error );
+		}
+	}
+	if ( window.onunhandledrejection ) {
+		for ( const rejection of rejectionQueue ) {
+			window.onunhandledrejection( rejection );
+		}
+	}
+}
+
+function beforeBreadcrumb( breadcrumb: SentryApi.Breadcrumb ): SentryApi.Breadcrumb | null {
+	// Ignore default navigation events -- we'll track them ourselves in the page( '*' ) handler.
+	if ( breadcrumb.category === 'navigation' && ! breadcrumb.data?.should_capture ) {
+		return null;
+	} else if ( breadcrumb.data?.should_capture ) {
+		delete breadcrumb.data.should_capture;
+	}
+	return breadcrumb;
+}
+
+interface SentryOptions {
+	beforeSend: ( e: SentryApi.Event ) => SentryApi.Event | null;
+}
+export async function initSentry( { beforeSend }: SentryOptions ) {
+	// Make sure we don't throw
+	try {
+		// No Sentry loading on the server.
+		// No double-loading.
+		if ( typeof document === 'undefined' || state.state !== 'initial' ) {
+			return;
+		}
+		state = { state: 'loading' };
+
+		// Enable Sentry only for 10% of requests or always in calypso.live for testing.
+		// Always disable if catch-js-errors is not available in the environment.
+		if (
+			! (
+				config.isEnabled( 'catch-js-errors' ) &&
+				( config( 'env_id' ) === 'wpcalypso' || Math.floor( Math.random() * 10 ) === 1 )
```

Comment:
Curious - where is the `env_id` set?

Just a nit and no need to act on that but, a better value for `calypso.live` would probably be `calypso.live`? :)

Reply:
Everything from `config` comes from all the normal config files here: https://github.com/Automattic/wp-calypso/tree/trunk/config. Basically different data for the different environments :) I agree that the name is confusing, but I think it's been this way for long enough that we might break things if we change it :(

Reply:
Thanks for pointing that out :)